Content.Server/Teleportation/TeleportLocationsSystem.cs

```diff
@@ -0,0 +1,58 @@
+﻿using Content.Server.Chat.Systems;
+using Content.Server.Warps;
+using Content.Shared.Teleportation;
+using Content.Shared.Teleportation.Components;
+using Content.Shared.Teleportation.Systems;
+
+namespace Content.Server.Teleportation;
+
+public sealed partial class TeleportLocationsSystem : SharedTeleportLocationsSystem
+{
+    [Dependency] private readonly ChatSystem _chat = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<TeleportLocationsComponent, MapInitEvent>(OnInit);
+        SubscribeLocalEvent<TeleportLocationsComponent, TeleportLocationRequestPointsMessage>(OnTeleportLocationRequest);
+    }
+
+    private void OnInit(Entity<TeleportLocationsComponent> ent, ref MapInitEvent args)
+    {
+        UpdateTeleportPoints(ent);
+    }
+
+    private void OnTeleportLocationRequest(Entity<TeleportLocationsComponent> ent, ref TeleportLocationRequestPointsMessage args)
+    {
+        UpdateTeleportPoints(ent);
+    }
+
+    /// <summary>
+    ///     Gets the teleport points to send to the BUI
+    /// </summary>
+    private void UpdateTeleportPoints(Entity<TeleportLocationsComponent> ent)
+    {
+        var allEnts = AllEntityQuery<WarpPointComponent>();
+
+        while (allEnts.MoveNext(out var warpEnt, out var warpPointComp))
+        {
+            if (warpPointComp.GhostOnly || string.IsNullOrWhiteSpace(warpPointComp.Location))
+                continue;
+
+            ent.Comp.AvailableWarps.Add(new TeleportPoint(warpPointComp.Location, GetNetEntity(warpEnt)));
+        }
+
+        Dirty(ent);
+    }
+
+    protected override void OnTeleportLocationRequest(Entity<TeleportLocationsComponent> ent, ref TeleportLocationRequestTeleportMessage args)
+    {
+        if (ent.Comp.User is null || string.IsNullOrWhiteSpace(ent.Comp.Speech))
+            return;
+
+        _chat.TrySendInGameICMessage(ent.Comp.User.Value, $"{ent.Comp.Speech.Trim()} ({args.PointName})", InGameICChatType.Speak, ChatTransmitRange.Normal);
+
+        base.OnTeleportLocationRequest(ent, ref args);
+    }
+}
```

Comment:
Isn't this just a copy of ghost warp, why don't we just use that UI again?

Reply:
It has player teleports baked into it pretty hard and if the ghost should follow the warp point.

Reply:
The main issue with this is it also couples it to wizard and doesn't support a subset of teleports either. The component name says it's teleportlocations but functionally it only supports that.

Reply:
Kind of? You could throw it on a console and basically give it to Abductors/Clock Cult with some modification.
Definitely not as freeform as just using an AI eye on a camera console and hitting teleport/recall but basically can still be used as part of a teleport console. The effect itself can be replaced with anything like sparks, a silhouette, etc and the speech itself is optional. In theory if you put it on a console you could proooobably make the console speak.

At the very least since this uses the BUI system as a backbone it has more functionality than the ghost menu.

What thoughts did you have?